### ISSUE TITLE
Stops functioning after unplugging external monitor

### ISSUE THREAD

**Issue by sidazhang** [koekeishiya/kwm#330]
I currently have a single external monitor setup. The external monitor plus the laptop screen makes 2 different sized screens.

When I unplug the external monitor, all the windows move to the laptop screen. That immediately breaks KWM, it stops functioning all together. It starts functioning again after I plug the external display back in.

**Comment by cHoco** [comment]
Would you mind trying this patch https://gist.github.com/cHoco/574a7e0d1e58aa021180665bb34260db/raw and see if solves the problem? In any case please post the debug logs, thank you !

**Comment by koekeishiya** [comment]
After the patch from cHoco, this is no longer an issue (I'm unable to reproduce at least).

**Comment by sidazhang** [comment]
I built with the latest master. I still encounter the issue.

My process:

```
git clone
make install
cp examples/com.koekeishiya.kwm.plist ~/Library/LaunchAgents/
./bin/kwm
```

Logs when I unplug the external display:

```
➜  kwm git:(master) ./bin/kwm
Apr 22 23:10:29  kwm[3747] <Error>: Set a breakpoint at CGSLogError to catch errors as they are logged.
Apr 22 23:10:29  kwm[3747] <Error>: CGSCopyDisplayUUID: Invalid display 0x0424eea9
Apr 22 23:11:06  kwm[3747] <Error>: CGSCopyDisplayUUID: Invalid display 0x0424eea9
```

Environment:
OSX 10.11.4 (15E65)
```
➜  ~  swiftc -v
Apple Swift version 2.2 (swiftlang-703.0.18.1 clang-703.0.29)
Target: x86_64-apple-macosx10.9
➜  ~  clang -v
Apple LLVM version 7.3.0 (clang-703.0.29)
Target: x86_64-apple-darwin15.4.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

**Comment by sidazhang** [comment]
@cHoco @koekeishiya One more thing, The external display is apple 27'' thunderbolt display

**Comment by cHoco** [comment]
Hi, sorry for if I wasn't clear, but I'll need the logs of the debug build, so you have to rebuild it without issuing `make install` just `make clean && make` Thank you :)

**Comment by sidazhang** [comment]
Debug logs:

```
➜  kwm git:(master) ./bin/kwm
Local Daemon is now running..
ChangeSplitRatio() New Split-Ratio is 0.5
Creating screen info for display ID: 69529257 Resolution: 2560x1440 Origin: (0,0)
DisplayID 69529257 has index 0 and Identifier 8A934838-3B40-D138-6FB7-E282238D2B56
Creating screen info for display ID: 69732992 Resolution: 1280x800 Origin: (638,1440)
DisplayID 69732992 has index 1 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
CreateWindowNodeTree() Create Space 2
CreateBSPTree() Create pair of leafs
CreateBSPTree() Create pair of leafs
SetWindowDimensions() Inbox (184) - sidazhang89@gmail.com - Gmail pos: 15,15 size: 1255,695
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 15,730 size: 1255,695
SetWindowDimensions() 1. sidazhang@sidazhang: ~/Uber/scripts/kwm (zsh) pos: 1293,15 size: 1252,1410
SetWindowDimensions() 1. sidazhang@sidazhang: ~/Uber/scripts/kwm (zsh) pos: 1293,15 size: 1252,1410

[0] BEGIN DISPLAY CONFIGURATION CALLBACK
[0] END DISPLAY CONFIGURATION CALLBACK


[1] BEGIN DISPLAY CONFIGURATION CALLBACK
[1] END DISPLAY CONFIGURATION CALLBACK


[2] BEGIN DISPLAY CONFIGURATION CALLBACK
Apr 23 09:58:33  kwm[19742] <Error>: Set a breakpoint at CGSLogError to catch errors as they are logged.
Apr 23 09:58:33  kwm[19742] <Error>: CGSCopyDisplayUUID: Invalid display 0x0424eea9
Display 69529257 is asleep!
DisplayID 69732992 has index 0 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
Display 69529257 disabled
[2] END DISPLAY CONFIGURATION CALLBACK


[3] BEGIN DISPLAY CONFIGURATION CALLBACK
Display 69732992 changed resolution
Display 69732992 became main
Display 69732992 changed mode
[3] END DISPLAY CONFIGURATION CALLBACK

GetWindowRef() Could not get AppWindowLst
SetWindowDimensions() 1. ./bin/kwm (kwm) pos: 1293,278 size: 1252,1147
GetWindowRef() Could not get AppWindowLst
SetWindowRefFocus() Focused Window: Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm 0,23
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 746,438
SetWindowRefFocus() Focused Window: Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm 0,23
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 746,438
SetWindowDimensions() 1. ./bin/kwm (kwm) pos: 1290,277 size: 1255,1147
SetWindowDimensions() 1. ./bin/kwm (kwm) (101,50) pos: 1290,277 size: 1255,1147

[4] BEGIN DISPLAY CONFIGURATION CALLBACK
[4] END DISPLAY CONFIGURATION CALLBACK


[5] BEGIN DISPLAY CONFIGURATION CALLBACK
[5] END DISPLAY CONFIGURATION CALLBACK


[6] BEGIN DISPLAY CONFIGURATION CALLBACK
The display that is being added is already there!
DisplayID 69529257 has index 0 and Identifier 8A934838-3B40-D138-6FB7-E282238D2B56
DisplayID 69732992 has index 1 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
Display 69529257 changed resolution
SetWindowDimensions() Inbox (184) - sidazhang89@gmail.com - Gmail pos: 15,15 size: 1255,695
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 15,730 size: 1255,695
SetWindowDimensions() 1. ./bin/kwm (kwm) pos: 1293,15 size: 1252,1410
Display 69529257 became main
Display 69529257 changed mode
Display 69529257 enabled
[6] END DISPLAY CONFIGURATION CALLBACK


[7] BEGIN DISPLAY CONFIGURATION CALLBACK
Display 69732992 changed resolution
Display 69732992 changed mode
[7] END DISPLAY CONFIGURATION CALLBACK

GetWindowRef() Could not get AppWindowLst
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 1293,23
UpdateActiveScreen() Active Display Changed
GiveFocusToScreen() 1: Space transition ended 2 -> 38
Uninitialized Screen
UpdateActiveScreen() Active Display Changed
GiveFocusToScreen() 0: Space transition ended 38 -> 2
Populated Screen Key/Mouse Focus
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 1293,23
```

The logs of the debug build itself
```
➜  kwm git:(master) make
g++ -c kwm/kwm.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/kwm.o
g++ -c kwm/container.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/container.o
g++ -c kwm/node.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/node.o
g++ -c kwm/tree.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/tree.o
g++ -c kwm/window.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/window.o
kwm/window.cpp:1483:5: warning: 'GetProcessForPID' is deprecated: first deprecated in OS X 10.9 [-Wdeprecated-declarations]
    GetProcessForPID(KWMFocus.Window->PID, &NewPSN);
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:930:1: note:
      'GetProcessForPID' has been explicitly marked deprecated here
GetProcessForPID(
^
kwm/window.cpp:1523:5: warning: 'GetProcessForPID' is deprecated: first deprecated in OS X 10.9 [-Wdeprecated-declarations]
    GetProcessForPID(KWMFocus.Window->PID, &NewPSN);
    ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:930:1: note:
      'GetProcessForPID' has been explicitly marked deprecated here
GetProcessForPID(
^
kwm/window.cpp:1531:9: warning: 'SetFrontProcessWithOptions' is deprecated: first deprecated in OS X 10.9 [-Wdeprecated-declarations]
        SetFrontProcessWithOptions(&KWMFocus.PSN, kSetFrontProcessFrontWindowOnly);
        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/System/Library/Frameworks/ApplicationServices.framework/Frameworks/HIServices.framework/Headers/Processes.h:658:1: note:
      'SetFrontProcessWithOptions' has been explicitly marked deprecated here
SetFrontProcessWithOptions(
^
3 warnings generated.
g++ -c kwm/application.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/application.o
g++ -c kwm/display.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/display.o
kwm/display.cpp:17:9: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
    for(auto it = KWMTiling.DisplayMap.begin(); it != KWMTiling.DisplayMap.end(); it++)
        ^
kwm/display.cpp:29:5: warning: 'auto' type specifier is a C++11 extension [-Wc++11-extensions]
    auto it = KWMTiling.DisplayMap.find(oldDisplayIdx);
    ^
2 warnings generated.
g++ -c kwm/daemon.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/daemon.o
g++ -c kwm/interpreter.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/interpreter.o
g++ -c kwm/keys.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/keys.o
g++ -c kwm/space.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/space.o
g++ -c kwm/border.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/border.o
g++ -c kwm/notifications.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/notifications.o
g++ -c kwm/workspace.mm -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/workspace.o
g++ -c kwm/serializer.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/serializer.o
g++ -c kwm/tokenizer.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/tokenizer.o
g++ -c kwm/rules.cpp -DDEBUG_BUILD -g -O3 -Wall -o obj/kwm/rules.o
mkdir -p ./bin && mkdir -p /Users/sidazhang/.kwm
g++ obj/kwm/kwm.o obj/kwm/container.o obj/kwm/node.o obj/kwm/tree.o obj/kwm/window.o obj/kwm/application.o obj/kwm/display.o obj/kwm/daemon.o obj/kwm/interpreter.o obj/kwm/keys.o obj/kwm/space.o obj/kwm/border.o obj/kwm/notifications.o obj/kwm/workspace.o obj/kwm/serializer.o obj/kwm/tokenizer.o obj/kwm/rules.o -DDEBUG_BUILD -g -O3 -Wall -lpthread -framework ApplicationServices -framework Carbon -framework Cocoa -o bin/kwm
g++ -c kwmc/kwmc.cpp -O3 -Wall -o obj/kwmc/kwmc.o
g++ obj/kwmc/kwmc.o -O3 -Wall -o bin/kwmc
swiftc -c kwm-overlay/kwm-overlay.swift -DDEBUG_BUILD -g -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk -o obj/kwm-overlay/kwm-overlay.o
swiftc obj/kwm-overlay/kwm-overlay.o -lc++ -L /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx -Xlinker -force_load_swift_libs -o bin/kwm-overlay
mkdir -p /Users/sidazhang/.kwm
if test ! -e /Users/sidazhang/.kwm/kwmrc; then cp -n examples/kwmrc /Users/sidazhang/.kwm/kwmrc; fi
```

**Comment by cHoco** [comment]
How do you disconnect the monitor? Because from the logs it seems that it is indeed removed, but it's in a sleep state.

**Comment by sidazhang** [comment]
@cHoco I unplug the thunderbolt cable from my laptop

**Comment by cHoco** [comment]
Ok, I tried doing the same with my tv and just like you the disconnected display is marked asleep, but for some reason my tree get rebalanced anyway. I need to first understand why is working on my setup (even though it shouldn't ?!?) and how to differentiate from a disconnected monitor and an asleep one (very doubtful there's a way....)

**Comment by koekeishiya** [comment]
>  differentiate from a disconnected monitor and an asleep one (very doubtful there's a way....)

The last time I worked on the display code, the display goes to sleep when it is disconnected. I wonder if the `disabled` flag would be worth checking out.

**Comment by cHoco** [comment]
You mean `kCGDisplayDisabledFlag` ?  I already log that in the callback and it's set both times :(

I still don't get why it's working for me or you ahah
I really don't know what would be the best way to solve this.... Maybe completely remove the sleep check since it seems to be useless. Instead of just destroying the node trees of the removed display we could maybe copy them in another structure `KWMTiling.RemovedDisplayMap`, remove them and then check for it when a new display is connected. When that happens we then try to restore it by removing windows that were added to the remaining display trees and add the newly connected display to the main `DisplayMap`

**Comment by sidazhang** [comment]
Is there anything that would be helpful for me to do right now to help you
gather more data?

For example, I could take a video of me unplugging the external monitor and
the state that the laptop goes into afterwards.

**Comment by cHoco** [comment]
Just to confirm the issue is where I think it is, could you apply this patch https://gist.github.com/cHoco/6ab3b2667b9f0114a72291bad91ffdc9/raw and see if it works?

I'll probably ask for testing when I code the patch with the logic I described above. 👍

**Comment by sidazhang** [comment]
@cHoco Yep! It worked!

```
➜  kwm git:(master) ✗ ./bin/kwm
Local Daemon is now running..
ChangeSplitRatio() New Split-Ratio is 0.5
Creating screen info for display ID: 69529258 Resolution: 2560x1440 Origin: (0,0)
DisplayID 69529258 has index 0 and Identifier 8A934838-3B40-D138-6FB7-E282238D2B56
Creating screen info for display ID: 69732992 Resolution: 1280x800 Origin: (638,1440)
DisplayID 69732992 has index 1 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
CreateWindowNodeTree() Create Space 3
CreateBSPTree() Create pair of leafs
CreateBSPTree() Create pair of leafs
SetWindowDimensions() Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail pos: 15,15 size: 1255,695
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 15,730 size: 1255,695
SetWindowDimensions() 1. sidazhang@sidazhang: ~/Uber/scripts/kwm (zsh) pos: 1293,15 size: 1252,1410
SetWindowDimensions() 1. sidazhang@sidazhang: ~/Uber/scripts/kwm (zsh) pos: 1293,15 size: 1252,1410

[0] BEGIN DISPLAY CONFIGURATION CALLBACK
[0] END DISPLAY CONFIGURATION CALLBACK


[1] BEGIN DISPLAY CONFIGURATION CALLBACK
[1] END DISPLAY CONFIGURATION CALLBACK


[2] BEGIN DISPLAY CONFIGURATION CALLBACK
Apr 24 12:19:28  kwm[45230] <Error>: Set a breakpoint at CGSLogError to catch errors as they are logged.
Apr 24 12:19:28  kwm[45230] <Error>: CGSCopyDisplayUUID: Invalid display 0x0424eeaa
Display has been removed! DisplayID: 69529258
DisplayID 69732992 has index 0 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
GiveFocusToScreen() 0: Space transition ended 3 -> 3
Uninitialized Screen
Display 69529258 disabled
[2] END DISPLAY CONFIGURATION CALLBACK

CreateWindowNodeTree() Create Space 3
CreateBSPTree() Create pair of leafs
CreateBSPTree() Create pair of leafs
SetWindowDimensions() Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail pos: 15,15 size: 695,425
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 15,460 size: 695,425
SetWindowDimensions() 1. ./bin/kwm (kwm) (201,50) pos: 733,16 size: 692,868

[3] BEGIN DISPLAY CONFIGURATION CALLBACK
Display 69732992 changed resolution
SetWindowDimensions() Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail pos: 15,15 size: 695,425
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 15,460 size: 695,425
SetWindowDimensions() 1. ./bin/kwm (kwm) (201,50) pos: 733,16 size: 692,868
Display 69732992 became main
Display 69732992 changed mode
[3] END DISPLAY CONFIGURATION CALLBACK

Restarting Event Tap
SetWindowRefFocus() Focused Window: Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm 15,460
SetWindowRefFocus() Focused Window: Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail 15,23
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 733,23
SetWindowRefFocus() Focused Window: Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm 15,460

[4] BEGIN DISPLAY CONFIGURATION CALLBACK
[4] END DISPLAY CONFIGURATION CALLBACK


[5] BEGIN DISPLAY CONFIGURATION CALLBACK
[5] END DISPLAY CONFIGURATION CALLBACK

UpdateActiveSpace() Space transition ended 3 -> 79
CreateWindowNodeTree() Create Space 79
CreateBSPTree() Create pair of leafs
CreateBSPTree() Create pair of leafs
SetWindowDimensions() Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail pos: 15,15 size: 695,425
SetWindowDimensions() 1. ./bin/kwm (kwm) pos: 18,464 size: 692,421
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 730,15 size: 695,870

[6] BEGIN DISPLAY CONFIGURATION CALLBACK
New display detected! DisplayID: 69529258 Index: 1
Creating screen info for display ID: 69529258 Resolution: 2560x1440 Origin: (0,0)
DisplayID 69529258 has index 0 and Identifier 8A934838-3B40-D138-6FB7-E282238D2B56
DisplayID 69732992 has index 1 and Identifier 0AA6E386-CE2C-5EE2-CC95-F2B67DCDF861
GiveFocusToScreen() 0: Space transition ended 79 -> 3
Uninitialized Screen
Display 69529258 changed resolution
Display 69529258 became main
Display 69529258 changed mode
Display 69529258 enabled
[6] END DISPLAY CONFIGURATION CALLBACK


[7] BEGIN DISPLAY CONFIGURATION CALLBACK
Display 69732992 changed resolution
SetWindowDimensions() Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail pos: 653,1455 size: 615,375
SetWindowDimensions() 1. ./bin/kwm (kwm) pos: 654,1854 size: 613,370
SetWindowDimensions() Stops functioning after unplugging external monitor · Issue #330 · koekeishiya/kwm pos: 1288,1455 size: 615,770
Display 69732992 changed mode
[7] END DISPLAY CONFIGURATION CALLBACK

CreateWindowNodeTree() Create Space 3
Restarting Event Tap
GetWindowRef() Could not get AppWindowLst
UpdateActiveScreen() Active Display Changed
GiveFocusToScreen() 1: Space transition ended 3 -> 79
Populated Screen Key/Mouse Focus
GetWindowRef() Could not get AppWindowLst
GetWindowRef() Could not get AppWindowLst
SetWindowRefFocus() Focused Window: Inbox (3) - sidney@sidazhang.com - Sidney Sida Zhang Mail 653,1463
SetWindowRefFocus() Focused Window: 1. ./bin/kwm (kwm) 654,1854
```

**Comment by cHoco** [comment]
Ok, then now I just have to work on a general solution that works for both you and @chrishoage :D
The idea I described above stands, it'll just need some work.

Ok now some ramblings about this issue, I wanna write them down as both *reference* and also because it may be useful so someone else working on similar issue/wanting to take this task on.
Continuing from where I left above, restoring as it is the displaymap doesn't work, because to correctly restore the spaces we'll need to also update their ids, since CoreGraphics assigns new space ids (not always...) But which old id corresponds to the new id? Fortunately it seems that each space has a UUID assigned that doesn't change on display connect/disconnect so we'll probably be able to use that :) (it can be extracted from the dictionaries obtained using `CGSCopyManagedDisplaySpaces`) The only space that doesn't have a UUID is the first one, but this fact makes it unique regardless. The hard part now is to decide when to assign to a `space_info` the UUID and when to check if it has changed. Space initialisation and transition should probably be enough. Not sure if I'll be able to work on this in the next couple of days, so if anyone wants to work on this I won't get offended ahah 😛

**Comment by koekeishiya** [comment]
>  how to differentiate from a disconnected monitor and an asleep one (very doubtful there's a way....)

There is a function `CGDisplayIsOnline ( CGDirectDisplayID display );`

> If true, the specified display is connected; otherwise, false.

> A display is considered connected or online when the framebuffer hardware is connected to a monitor.
You can use this function to determine whether someone has plugged a display into the system while the main power was on. This hardware feature, called hot-plugging, may not be present on all displays.

I suspect you could also check the difference of the display lists returned by `CGGetOnlineDisplayList` and `CGGetActiveDisplayList`.

**Comment by koekeishiya** [comment]
I suspect this issue to be fixed in v3.0.6.
Feel free to reopen or create a new if that is not the case.

**Comment by sidazhang** [comment]
@koekeishiya Can you reopen this issue

KWM goes crazy after I unplug the external monitor. I am currently using v3.0.6

![2016-08-13 18 05 17](https://cloud.githubusercontent.com/assets/3519115/17646496/e7b79c0e-6180-11e6-9f38-eabc65da1c2f.gif)

**Comment by sidazhang** [comment]
Not only that it goes a bit crazy, it also does not resize correctly to the new window width. You can see that I have 2 chrome windows. The left window occupies 80% of the width and the right table is pushed outside the of the screen altogether

**Comment by koekeishiya** [comment]
Could you clone master and run with a debug build, and post the output ? The information posted earlier doesn't really apply anymore  after v3.

**Comment by sidazhang** [comment]
This is fixed after upgrading to 4.0.2
